### PR TITLE
Check CubeMX code generation in CI

### LINKED ISSUE #1687: Add CI check for CubeMX regeneration from ioc file
### Description of the task
We currently have an ioc file, which is the project file for the ST CubeMX, in `//firmware_new/boards/frankie_v1`. Anyone can change that file and forget to stage it, and we currently have no way of detecting that.

We need to explicitly call:

`bazel run --cpu=stm32h7 //firmware_new/tools:cubemx_regen firmware_new/boards/frankie_v1` to make sure the `ioc` file is not corrupt, and after the code generates, we need to make sure that it builds. 

### Acceptance criteria

- [ ] Add `cubemx_regen` to board tests in CI
- [ ] Make sure CI fails if corrupt `ioc` files are picked up by CubeMX
- [ ] Make sure CI fails if the code that was generated != code committed (see note)

Note: The last one is a bit tricky because we need to run `./fix_formatting.sh` after `cubemx_regen` to verify if the code was generated.

### Blocked By

N/A

### PR DESCRIPTION
**Pull request by xyztiger**
### Description

Added a script to be run during CI to check for corruption in the ioc file for CubeMX

### Testing Done

CI passes without a corrupt file;
currently testing to see if CI fails with a corrupt file

### Resolved Issues

resolves #1687 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: .travis/check_cubemx_ci.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+cd src
+bazel run --cpu=stm32h7 //firmware_new/tools:cubemx_regen firmware_new/boards/frankie_v1
+if [[ "$?" != 0 ]]; then
+    echo "There was a problem running the formatting script, stopping now."
+    exit 1
+fi
+
+cd ..
+# Fix formatting
+./formatting_scripts/fix_formatting.sh
```
Comment (akhilveeraghanta) [inline .travis/check_cubemx_ci.sh]: you want to run `git diff` after this command to see if the code was generated correctly

we care if there was a diff after `fix_formatting.sh`, the output of `fix_formatting.sh` itself we don't really care about

File: src/firmware_new/boards/frankie_v1/frankie_v1.ioc [hunk 2]
```diff
@@ -1,6 +1,8 @@
 #MicroXplorer Configuration settings - do not modify
-LWIP0.BSP.mode=
+<<<<<<< HEAD
+#LWIP0.BSP.mode=
 FREERTOS.Queues01=RobotLogProtoQ,16,TbotsProto_RobotLog,0,Dynamic,NULL,NULL
+>>>>>>> d3cb8535b42f4e950b7445c66243ea112370a42a
```
Comment (akhilveeraghanta) [inline src/firmware_new/boards/frankie_v1/frankie_v1.ioc]: this actually isn't an issue with your script, looks like cubemx doesn't give a crap and just runs even though the file is completely botched

try changing the numbers in the script to see if it fails

File: .travis/check_cubemx_ci.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+CURR_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+echo "=========================================================="
+echo " THIS SHOULD ONLY BE RUN IN CI, PLEASE DO NOT USE LOCALLY"
+echo "=========================================================="
+
+cd src
+bazel run --cpu=stm32h7 //firmware_new/tools:cubemx_regen firmware_new/boards/frankie_v1
+if [[ "$?" != 0 ]]; then
+    echo "There was a problem running the formatting script, stopping now."
```
Comment (akhilveeraghanta) [inline .travis/check_cubemx_ci.sh]: need to update this echo

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]

File: .travis/check_cubemx_ci.sh (hunk 1 above)
Comment (akhilveeraghanta) [reply .travis/check_cubemx_ci.sh]: git diff exits with 0 if there is no diff, so you just have to flip the check after

File: .travis.yml [hunk 4]
```diff
@@ -6,6 +6,7 @@ language: generic
 
 install:
     - ./environment_setup/setup_software.sh
+    - ./environment_setup/setup_firmware.sh
```
Comment (jonathanlew) [inline .travis.yml]: Can we do this only for the cubemx check?

File: .travis/check_cubemx_ci.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+bazel run --cpu=stm32h7 //firmware_new/tools:cubemx_regen firmware_new/boards/frankie_v1
+if [[ "$?" != 0 ]]; then
+    echo "There was an regenerating cubemx, stopping now."
+    exit 1
+fi
+
+cd ..
+# Fix formatting
+./formatting_scripts/fix_formatting.sh
+git diff
```
Comment (jonathanlew) [inline .travis/check_cubemx_ci.sh]:
```suggestion
git diff --exit-code
```
See: https://git-scm.com/docs/git-diff

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/firmware_new/boards/frankie_v1/ethernetif.c [hunk 6]
```diff
@@ -440,7 +440,7 @@ static void low_level_init(struct netif *netif)
  *
  * @note Returning ERR_MEM here if a DMA queue of your MAC is full can lead to
  *       strange results. You might consider waiting for space in the DMA queue
- *       to become available since the stack doesn't retry to send a packet
+ *       to become availale since the stack doesn't retry to send a packet
```
Comment (jonathanlew) [inline src/firmware_new/boards/frankie_v1/ethernetif.c:443]: ?

**Review by jonathanlew — COMMENTED** [review]
what does the error message look like when it fails?

**Comment by jonathanlew** [comment]
Also please change the title

File: src/firmware_new/boards/frankie_v1/ethernetif.c (hunk 6 above)
Comment (akhilveeraghanta) [reply src/firmware_new/boards/frankie_v1/ethernetif.c:443]: as stupid as this sounds, this gets autogenerated, and does not get fixed by codespell. So there is nothing we can do about this

**Comment by akhilveeraghanta** [comment]
> what does the error message look like when it fails?

it shows a diff of the files that were changed, but mostly if the check fails you should re-run regen and commit the auto generated changes

**Comment by jonathanlew** [comment]
> > what does the error message look like when it fails?
> 
> it shows a diff of the files that were changed, but mostly if the check fails you should re-run regen and commit the auto generated changes

can you paste an example?

**Comment by akhilveeraghanta** [comment]
> > > what does the error message look like when it fails?
> > 
> > 
> > it shows a diff of the files that were changed, but mostly if the check fails you should re-run regen and commit the auto generated changes
> 
> can you paste an example?

```
 Code generated is different from code committed, stopping now.
Error: Process completed with exit code 1.
```
https://github.com/UBC-Thunderbots/Software/pull/1816/commits/3e5da6a45bc4ee21a0e34e22d63352e6f9510899
https://github.com/UBC-Thunderbots/Software/pull/1816/checks?check_run_id=1999128223

**Review by jonathanlew — APPROVED** [review]
LGTM! :+1:

**Review by HaomiaoZ — APPROVED** [review]
LGTM 👍

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**